### PR TITLE
SSP Cache Implementation

### LINKED ISSUE #254: Redundant text "v diagramu"
The label "v diagramu" seems redundant
image

### PR DESCRIPTION
**Pull request by bindeali**
## PR Info:
* Can be merged and closed right away.
* Closes #254.
* Implemented the SSP Cache.

## User guide:
* The search function now also searches the SSP cache and displays results below the usual workspace vocabularies. Hovering over a term reveals additional information about the term.

https://user-images.githubusercontent.com/56399637/116725116-d28b1c80-a9e1-11eb-8b04-095eea81b814.mp4
* The 📚 button can be used to group results by vocabularies.

https://user-images.githubusercontent.com/56399637/116725640-7d9bd600-a9e2-11eb-8551-d8a997db6bbd.mp4

* The select can be used to filter results by vocabulary or vocabularies. If a vocabulary is selected and the search field is empty, the whole vocabulary is shown.

https://user-images.githubusercontent.com/56399637/116727166-7e356c00-a9e4-11eb-8b78-654e66b45512.mp4

* A ⭐ symbol next to the term signifies that the term is already in the workspace.

https://user-images.githubusercontent.com/56399637/116727249-9907e080-a9e4-11eb-808a-05e81e9146d9.mp4

* To add a term to the workspace, drag it onto the canvas as with any other term. The terms added can be read only. You can also ctrl + click onto terms to select multiple terms and ctrl + click on vocabularies to select all terms in the group. 

https://user-images.githubusercontent.com/56399637/116727274-9f965800-a9e4-11eb-9110-600310100207.mp4

* The terms are added under a new read-only vocabulary. A counter is displayed if the vocabulary is not fully represented in the workspace. Clicking on this counter displays all the terms of the vocabulary.

https://user-images.githubusercontent.com/56399637/116727287-a3c27580-a9e4-11eb-954e-3a2aa5dd8ad3.mp4

## Minor changes:
* ontoGrapher now saves data for a workspace into workspace-specific contexts. This requires that existing workspace data be moved into this context. **Please note that moving read-only vocabularies is unsupported. The old data is not deleted.** (I noticed that some workspaces still had read-only vocabularies in the data despite only read/write vocabularies present in Control Panel. These vocabularies are not brought over).
ezgif-2-6e3bb5fd7088
* Users are now notified whenever they open a new version of ontoGrapher.
* Some refactoring.

### **Unfortunately the deploy previews are currently unavailable due to an issue with the `slovník-test` server.**

* [Veřejné rejstříky](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-224258741)
* [eSbírka](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1326236660)
* [Datový model OFN číselníků a statistických klasifikací](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-2083940323)
* [Karel test](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-415314821)
* [AB test](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance2011031447)
* [stenchlak-test](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1654748901)
* [Turisticky cil](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance1044794056)
* [čas](https://deploy-preview-266--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-202167279)

### CODE REVIEW

**Review by blcham — COMMENTED** [review]
Looks great !

I was going to test the functionality in netlify preview, but I found out it is not possible:
image

Are u using this build anyhow?

**Comment by bindeali** [comment]
Unfortunately the netlify deploy previews are not available due to CORS issues with `slovník-test`'s graphDB. It seems as if the problem is somewhere between the deploy configuration of netlify and/or `slovník-test` as testing on a copy of the DB works as intended. The issue is being investigated.

File: src/interface/ContextInterface.ts [hunk 1]
```diff
@@ -1,64 +1,127 @@
...
+    vocabularyQ
+  )
+    .then((response) => response.json())
+    .then((data) => {
+      return data.results.bindings;
+    })
+    .catch(() => false);
+  if (responseInit.length === 0) return false;
+  let vocabularies: {
+    [key: string]: {
```
Comment (psiotwo) [inline src/interface/ContextInterface.ts:65]: would suggest to use some types instead of maps of maps.

**Comment by psiotwo** [comment]
Functionally-wise, looks perfect - can't wait seeing it in dev :-)

File: src/function/FunctionDraw.ts [hunk 2]
```diff
@@ -1,98 +1,157 @@
...
+              4,
+            text.length > 0
+              ? 8 *
+                  text.reduce((a, b) => (a.length > b.length ? a : b), "")
+                    .length
+              : 0
+          )
+        : labels.reduce((a, b) => (a.length > b.length ? a : b), "").length *
+            10 +
+          4;
```
Comment (karelklima) [inline src/function/FunctionDraw.ts:97]: A ternary operator as a function argument, within a ternary operator... not readable at all. The reducer function is redundant.

File: src/function/FunctionEditVars.ts [hunk 3]
```diff
@@ -105,104 +109,159 @@ export function parsePrefix(prefix: string, name: string): string {
...
+        );
     }
-    targets.forEach(target => {
-        const elem = Object.keys(ProjectElements).find(elem => ProjectElements[elem].connections.includes(target));
-        if (elem) ProjectElements[elem].connections.splice(ProjectElements[elem].connections.indexOf(target), 1);
-    })
-    VocabularyElements[ProjectElements[id].iri].labels = initLanguageObject("");
-    ProjectElements[id].connections = [];
+    targets.forEach((target) => {
+        const elem = Object.keys(WorkspaceElements).find((elem) =>
```
Comment (karelklima) [inline src/function/FunctionEditVars.ts:144]: Doesn't seem like this file went through Prettier

**Review by karelklima — APPROVED** [review]
I marked this as approved, but please note that PRs like this one are impossible to review (or at the very least quite difficult). It is unfortunate that there are formatting changes as well as functional changes present.

**Pull request merged**